README.md
[![arXiv](https://img.shields.io/badge/arXiv-1704.05694%20-red)](https://arxiv.org/abs/1704.05694)


NEURAL: A Neonatal EEG Feature Set in Matlab 
============================================

Matlab code to generate a set of quantitative features from multichannel EEG
recordings. Features include amplitude measures, spectral measures, and basic connectivity
measures (across hemisphere's only). Also, for preterm EEG (assuming gestational age < 32
weeks), will generate features from bursts annotations (e.g. maximum inter-burst
interval). Burst annotations require a separate package, also available on
[github](https://github.com/otoolej/burst_detector). 

Releases archived at Zenodo:
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.1052810.svg)](https://doi.org/10.5281/zenodo.1052810)


Full details of the methods are in:  

`JM O’Toole and GB Boylan (2017). NEURAL:
quantitative features for newborn EEG using Matlab. ArXiv e-prints, arXiv:1704.05694`  

which is available at https://arxiv.org/abs/1704.05694.

---

[Requirements](#requirements) | [Use](#use) | [Quantitative
features](#quantitative-features) | [Files](#files) | [Test computer
setup](#test-computer-setup) | [Licence](#licence) | [References](#references) |
[Contact](#contact)

## Requirements
Matlab (R2013 or newer,
[Mathworks](http://www.mathworks.co.uk/products/matlab/)) with the signal
processing toolbox and statistics toolbox. Not tested with Octave but should work with minor
tweaking.

## Use 

Set paths in Matlab, or do so using the `load_curdir` function:
```matlab
  >> load_curdir;
```

As an example, generate simulated EEG and calculate relative spectral power, standard
deviation of range-EEG, and brain symmetry index:
```matlab
	% generate EEG-like data (coloured Gaussian noise)
	data_st=gen_test_EEGdata(5*60,64,1);

	% define feature set (or can define in neural_parameters.m):
	feature_set={'spectral_relative_power','rEEG_SD', 'connectivity_BSI'};
	
	% estimate features:
	feat_st=generate_all_features(data_st,[],feature_set);
```

See the `demos/` directory for further examples. All parameters are set the file
`neural_parameters.m`.


## Quantitative features

The feature set contains amplitude, spectral, connectivity, and burst annotation features.
Amplitude features include range-EEG (D. O’ Reilly et al., 2012;
see [references](#references)), a clearly-defined alternative to amplitude-integrated EEG
(aEEG). All features are generated for four different frequency bands (typically 0.5–4,
4–7, 7–13, and 13–30 Hz), with some exceptions. The following table describes the features
in more detail:

| feature name               | description                                                                   | FB  |
|----------------------------|-------------------------------------------------------------------------------|-----|
| spectral\_power            | spectral power: absolute                                                      | yes |
| spectral\_relative\_power  | spectral power: relative (normalised to total spectral power)                 | yes |
| spectral\_flatness         | spectral entropy: Wiener (measure of spectral flatness)                       | yes |
| spectral\_entropy          | spectral entropy: Shannon                                                     | yes |
| spectral\_diff             | difference between consecutive short-time spectral estimates                  | yes |
| spectral\_edge\_frequency  | cut-off frequency (fc): 95% of spectral power contained between 0.5 and fc Hz | no  |
| FD                         | fractal dimension                                                             | yes |
| amplitude\_total\_power    | time-domain signal: total power                                               | yes |
| amplitude\_SD              | time-domain signal: standard deviation                                        | yes |
| amplitude\_skew            | time-domain signal: skewness (absolute value)                                 | yes |
| amplitude\_kurtosis        | time-domain signal: kurtosis                                                  | yes |
| amplitude\_env\_mean       | envelope: mean value                                                          | yes |
| amplitude\_env\_SD         | envelope: standard deviation (SD)                                             | yes |
| connectivity\_BSI          | brain symmetry index (see Van Putten 2007)                                    | yes |
| connectivity\_corr         | correlation (Spearman) between envelopes of hemisphere-paired channels        | yes |
| connectivity\_coh\_mean    | coherence: mean value                                                         | yes |
| connectivity\_coh\_max     | coherence: maximum value                                                      | yes |
| connectivity\_coh\_freqmax | coherence: frequency of maximum value                                         | yes |
| rEEG\_mean                 | range EEG: mean                                                               | yes |
| rEEG\_median               | range EEG: median                                                             | yes |
| rEEG\_lower\_margin        | range EEG: lower margin (5th percentile)                                      | yes |
| rEEG\_upper\_margin        | range EEG: upper margin (95th percentile)                                     | yes |
| rEEG\_width                | range EEG: upper margin - lower margin                                        | yes |
| rEEG\_SD                   | range EEG: standard deviation                                                 | yes |
| rEEG\_CV                   | range EEG: coefficient of variation                                           | yes |
| rEEG\_asymmetry            | range EEG: measure of skew about median                                       | yes |
| IBI\_length\_max           | burst annotation: maximum (95th percentile) inter-burst interval              | no  |
| IBI\_length\_median        | burst annotation: median inter-burst interval                                 | no  |
| IBI\_burst\_prc            | burst annotation: burst percentage                                            | no  |
| IBI\_burst\_number         | burst annotation: number of bursts                                            | no  |

FB: features generated for each frequency band (FB)


## Files
Some Matlab files (.m files) have a description and an example in the header. To read this
header, type `help <filename.m>` in Matlab.  Directory structure is as follows: 
```
├── amplitude_features/                  # amplitude features
├── spectral_features/                   # spectral features
├── connectivity_features/               # hemisphere connectivity features
├── range_EEG/                           # range-EEG (similar to aEEG)
├── IBI_features/                        # features from the burst annotations 
├── preprocessing/                       # loads EEG from EDF files (including artefact removal)
└── utils/                               # misc. functions
```
with some files of interest:
```
├── neural_parameters.m                  # all parameters defined here
├── all_features_list.m                  # complete list of functions (do not edit)
└── generate_all_features.m              # main function: generates feature set on EEG
```



## Test computer setup
- hardware:  Intel Core i7-8700K @ 4.5GHz; 32GB memory.
- operating system: Ubuntu GNU/Linux x86_64 distribution (19.10) with Linux kernel
  5.3.0-24-generic
- software: Matlab (R2019a)


## Licence

```
Copyright (c) 2016, John M. O' Toole, University College Cork
All rights reserved.

Redistribution and use in source and binary forms, with or without modification,
are permitted provided that the following conditions are met:

  Redistributions of source code must retain the above copyright notice, this
  list of conditions and the following disclaimer.

  Redistributions in binary form must reproduce the above copyright notice, this
  list of conditions and the following disclaimer in the documentation and/or
  other materials provided with the distribution.

  Neither the name of the University College Cork nor the names of its
  contributors may be used to endorse or promote products derived from
  this software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR
ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
(INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
```


## References

1. JM O’Toole and GB Boylan (2017). NEURAL: quantitative features for newborn EEG using
Matlab. ArXiv e-prints, arXiv:[1704.05694](https://arxiv.org/abs/1704.05694).

2. D O’Reilly, MA Navakatikyan, M Filip, D Greene, & LJ Van Marter (2012). Peak-to-peak
amplitude in neonatal brain monitoring of premature infants. Clinical Neurophysiology,
123(11):2139–2153.

3. MJAM van Putten (2007). The revised brain symmetry index. Clinical Neurophysiology,
118(11):2362–2367.

4. T Higuchi (1988). Approach to an irregular time series on the basis of the fractal theory,
Physica D: Nonlinear Phenomena, 31:277–283.

5. MJ Katz (1988). Fractals and the analysis of waveforms. Computers in Biology and
Medicine, 18(3):145–156.

6. AV Oppenheim, RW Schafer. Discrete-Time Signal Processing. Prentice-Hall, Englewood
Cliffs, NJ 07458, 1999.

7. JM O’ Toole, GB Boylan, S Vanhatalo, NJ Stevenson (2016). Estimating functional brain
maturity in very and extremely preterm neonates using automated analysis of the
electroencephalogram. Clinical Neurophysiology, 127(8):2910–2918

8. JM O’ Toole, GB Boylan, RO Lloyd, RM Goulding, S Vanhatalo, NJ Stevenson
    (2017). Detecting Bursts in the EEG of Very and Extremely Premature Infants using a
    Multi-Feature Approach. Medical Engineering and
    Physics, vol. 45, pp. 42-50, 2017. 
	[DOI:10.1016/j.medengphy.2017.04.003](https://dx.doi.org/10.1016/j.medengphy.2017.04.003)
	
9. JM O'Toole and GB Geraldine (2019). Quantitative Preterm EEG Analysis: The Need for
   Caution in Using Modern Data Science Techniques. Frontiers in Pediatrics 7, 174
   [DOI:10.3389/fped.2019.00174](https://doi.org/10.3389/fped.2019.00174)
	

## Contact

John M. O'Toole

Neonatal Brain Research Group,  
[INFANT Research Centre](https://www.infantcentre.ie/),  
Department of Paediatrics and Child Health,  
Room 2.19 UCC Academic Paediatric Unit, Cork University Hospital,  
University College Cork,  
Ireland

- email: jotoole AT ucc _dot ie 



CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. Project started
2016-04-01 but this log started 2017-03-16, after version 0.2.2. More details can be found
in the git logs.

## unreleased
### Changed 
### Removed
### Fixed
### Added


## [0.4.4] - 26-08-2020
### Changed 
- Returns Table of feature values (per epoch and per channel) from function
  `generate_all_features` if input argument `return_feat_epoch=true`
### Removed
### Fixed
- Bug fix for `connectivity_features`; was: if channel labels (cell of strings) are empty
  then would assume (incorrectly) only 2 channels and would use index [1,2] for left,right
  hemisphere; now: if no channel labels throw warning
### Added
- Check in `spectral_features`: is `L_window` longer than signal (and if so throw
  warning)?


## [0.4.3] - 2020-04-22
### Changed 
- changed behaviour for `IBI_burst_prc` and `IBI_burst_number` features: now removing any
  NaNs before calculating this metric
### Removed
### Fixed
- bug in rand_phase.m when taking the inverse-FFT (random signal will time-reversed but
  with same initial x[0] value); won't affect the coherence values.
- for rectangular window in `filter_zerophase.m` don't divide by window length
- for IBI features fixed removal of minimal interval of 16 samples and set to 1/4 seconds
  instead
- avoid error if maximum/median IBI duration is zero
- bug fix for `spectral_diff` feature: extra zero was included in spectral difference measure
  before taking the median
### Added

## [0.4.2] - 2019-04-29
### Changed 
- connectivity analysis: remove channel pairs that have common electrode, e.g. 'C3-Cz' and
  'Cz-C4'
### Removed
### Fixed
- tested with Matlab R2019a
- fixed syntax problem for fir1 function for R2019a
### Added


## [0.4.1] - 2019-03-19
### Changed 
- minor tweak to the overlap-and-add analysis when generating features on an epoch of EEG;
  now include as much of the signal as possible and pad with NaNs when necessary.
### Removed
### Fixed
- tested with Matlab R2018b;
- bug when downsampling: always using the resample function even when not necessary;
- bug in PSD power feature if length-epoch < (2 * PSD-window-length); 
- bug when length-epoch = signal-length, was summing over frequency bands.
### Added


## [0.4.0] - 2019-01-09
### Changed 
- Syntax for specifying the zero-level threshold of the coherence function in
  neural_parameters.m. NB: this will break compatibility with previous versions.
### Removed
- 'robust-PSD' for coherence function (as does not satisfy the Schwartz inequality).
### Fixed
- Coherence function for surrogate analysis, which assumed that auto-PSDs Pxx and Pyy
  would not change when randomising the phase of the Fourier transform for the surrogate
  signals. This holds only when PSD is a periodogram. Assumption removed and now Pxx and
  Pyy generated for each iteration.
### Added
- Analytic zero-coherence threshold (Halliday et al. 1995) for coherence
  estimates. Estimates coherence using the Bartlett PSD.


## [0.3.4] - 2018-09-18
### Changed 
- Higuchi method for fractal dimension can now incorporate NaNs
### Removed
### Fixed
- tested with Matlab R2018a 
### Added

## [0.3.3] - 2017-11-16
### Changed 
### Removed
### Fixed
- tested with Matlab R2017a
### Added
- licence file
- DOI links from zenodo


## [0.3.2] - 2017-06-16
### Changed 
### Removed
### Fixed
- reduced order of high-pass butterworth filter (with cut-off of 0.1Hz; see change in
  v0.3.1) from 5 to 2 in remove_artefacts.m function
### Added
- name of .mat file as input argument to 'resample_savemat.m'
- option in filter_butterworth_withnans.m to have separate filter orders for high and
  low-pass components of a bandpass filter
- file to plot rEEG (plot_rEEG.m)


## [0.3.1] - 2017-04-07
### Changed 
- bandpass filter changed from 0.5-40 Hz to 0.1-40 Hz before removing large-amplitude
  artefacts

### Removed
- bandpass filtering (0.5-40 Hz) on EEG prior to removing artefacts (constant values or
  sudden jumps in amplitude)

### Fixed
- avoid failing if only 1 channel for connectivity features (return NaN instead)
### Added


## [0.3.0] - 2017-03-29
### Changed 
- absolute and relative spectral power ('spectral\_power' and 'spectral\_relative\_power')
  now using periodogram (rectangular window) instead of Welch's power-spectral-density
  (PSD) estimate.
- all spectral estimates (PSD, robust-PSD, and periodogram) now used FFT length equal to
  window length (previously was in 2^n form, and at least >=256), i.e. no zero-padding.

### Removed
- PSD option for absolute and relative spectral power removed (replaced with periodogram).

### Fixed
- coherence function broken
- split of frequency bands for spectral and connectivity features when length of PSD is
  odd

### Added
- documentation to file headers
- this CHANGELOG.md!
- option to select PSD estimate for spectral features: either Welch PSD, robust-PSD
  (modified Welch PSD), or periodogram
- coherence function can use either PSD or robust-PSD 


## [0.2.2] - 2017-03-09
### Added
- surrogate data approach with coherence function
- NaN option to 'FILTER\_REPLACE\_ARTEFACTS'
- missing 'get\_window.m' and 'make\_odd.m' function
- now using semantic versioning (see http://semver.org/)

### Fixed
- scaling factor for robust-PSD



generate_all_features.m
%-------------------------------------------------------------------------------
% generate_all_features: generate all features for EEG recording (in .mat)
%
% Syntax: feat_st=generate_all_features(fname,channel_names,feat_set)
%
% Inputs: 
%     fname             - either EEG filename or data structure with EEG 
%                         e.g. data structure: data_st = gen_test_EEGdata(5*60,64,1); 
%     channel_names     - channel labels to process (default, process all)
%                         e.g. {'C3-O1','C4-O2','F3-C3','F4-C4'}
%     feat_set          - cell of features to compute, 
%                         e.g. {'spectral_relative_power','rEEG_SD', 'connectivity_BSI'}
%     return_feat_epoch - return features for each epoch? (true/false; default=false)
%                         (if true then returns 2nd output argument, 'feats_per_epoch')
% 
% Outputs: 
%     feat_st         - structure containing features
%     feats_per_epoch - table of features per epoch; in long format, with columns: 
%                         [time - channel - freq. band - value - feature name]
%                       start_time_sec: start time of the epoch (seconds)
%                       channel: name of channel (string)
%                       freq_band: number of frequency band (string, e.g. FB1)
%                       feature_value: value of the feature (real number, scalar)
%                       feature: name of the feature (string)         
%
% 
% Example:
%       % generate 5 minutes of simulated multichannel EEG data, with 64 Hz sample frequency
%       data_st=gen_test_EEGdata(5*60,64,1);
%
%       % select features to compute:
%       feature_set={'spectral_relative_power','rEEG_SD', 'connectivity_BSI'};
%
%       % generate all features:
%       feat_st=generate_all_features(data_st,[],feature_set);
% 

% John M. O' Toole, University College Cork
% Started: 07-04-2016
%
% last update: Time-stamp: <2020-08-17 18:08:33 (otoolej)>
%-------------------------------------------------------------------------------
function [feat_st, feats_all_epochs_tb] = generate_all_features(fname, channel_names, feat_set, ...
                                                      return_feat_epoch)
if(nargin<2 || isempty(channel_names)), channel_names=[]; end
if(nargin<3 || isempty(feat_set)), feat_set=[]; end
if(nargin<4 || isempty(return_feat_epoch)), return_feat_epoch = false; end



neural_parameters;


%---------------------------------------------------------------------
% 1. load EEG data from .mat file
%---------------------------------------------------------------------
if(isstruct(fname))
    % can input data as structure:
    eeg_data=fname.eeg_data; Fs=fname.Fs; ch_labels=fname.ch_labels;
    
else
    if(length(fname)>4 && strcmp(fname(end-3:end),'.mat'))
        fname=fname(1:end-4);
    end

    % load from .mat file:
    d=load([EEG_DATA_DIR_MATFILES fname '.mat']);
    fprintf(col_str(' loading EEG data from file saved on %s\n',1),datestr(d.time_now));
    eeg_data=d.eeg_data; Fs=d.Fs; ch_labels=d.ch_labels;

end
feats_per_epochs=[];


% select channels:
if(~isempty(channel_names))
    ikeep=[];
    for n=1:length(channel_names)
        it=find( strcmp(ch_labels,channel_names{n}) );
        if(~isempty(it))
            ikeep=[ikeep it];
        end
    end
    eeg_data=eeg_data(ikeep,:);
    ch_labels=ch_labels(ikeep);
end

% or remove empty channels:
irem=[];
for n=1:length(ch_labels)
    if( all(isnan(eeg_data(n,:))) )
        irem=[irem n];
    end
end
if(~isempty(irem))
% $$$     fprintf('removing channels: %s\n',ch_labels(irem));
    eeg_data(irem,:)=[]; ch_labels(irem)=[];
end


[N_channels,N]=size(eeg_data);

%---------------------------------------------------------------------
% 2. generate features
%---------------------------------------------------------------------
if(isempty(feat_set))
    feat_set=FEATURE_SET_ALL;
end

N_feats=length(feat_set);
feats_all_epochs_tb = [];

% A) iterate over features
for n=1:N_feats

    L_feature=size_feature(feat_set{n});
    feat_group=strsplit(feat_set{n},'_');
    feat_group=feat_group{1};

    %---------------------------------------------------------------------
    % SPECTRAL and AMPLITUDE
    % (analysis on a per-channel basis and divide each channel into epochs)
    %---------------------------------------------------------------------
    if( any(strcmp({'amplitude','spectral','rEEG','FD'},feat_group)) )

        % B) iterate over channels
        feats_channel=[]; x_epochs=[]; feats_tbl = [];
        for c=1:N_channels
            [x_epochs, epoch_start_times] = overlap_epochs(eeg_data(c,:)',Fs,EPOCH_LENGTH,EPOCH_OVERLAP);
            N_epochs=size(x_epochs,1);
            
            % C) iterate over epochs
            feats_epochs=NaN(N_epochs,L_feature);
            for e=1:N_epochs
                L_nans=length(find(isnan(x_epochs(e,:))));
                
                if(100*(L_nans/length(x_epochs(e,:))) < EPOCH_IGNORE_PRC_NANS)
                    if(strcmp(feat_group,'spectral'))
                        feats_epochs(e,:)=spectral_features(x_epochs(e,:),Fs, ...
                                                            feat_set{n});
                        
                    elseif(strcmp(feat_group,'FD'))
                        feats_epochs(e,:)=fd_features(x_epochs(e,:),Fs);
                        
                    elseif(strcmp(feat_group,'amplitude'))
                        feats_epochs(e,:)=amplitude_features(x_epochs(e,:),Fs, ...
                                                             feat_set{n});
                        
                    elseif(strcmp(feat_group,'rEEG'))
                        feats_epochs(e,:)=rEEG(x_epochs(e,:),Fs,feat_set{n});
                        
                    end
                end
            end
            % if want to return feature estimated over all epochs:
            if(return_feat_epoch)
        	% feats_per_epochs{n}(c,:,:)=feats_epochs;
                
                % create table with features and start time of epoch:
                fb_names = arrayfun(@(x) ['FB' num2str(x)], 1:size(feats_epochs, 2), 'un', false);
                tb = array2table([epoch_start_times' feats_epochs], ...
                                 'VariableNames', ['start_time_sec', fb_names]);
                % add channel:
                tb.channel(:) = string(ch_labels{c});
                
                % convert from wide to long format for frequency bands:
                tb = stack(tb, fb_names, 'newDataVariableName', {'feature_value'}, ...
                           'IndexVariableName', {'freq_band'});
                
                feats_tbl = [feats_tbl; tb];
            end
            
            % median over all epochs
            feats_channel(c,:)=nanmedian(feats_epochs, 1);
        end
        % and median over all channels:
        feat_st.(char(feat_set{n}))=nanmedian(feats_channel, 1);
        

        if(return_feat_epoch)
            % add feature name and combine:
            feats_tbl.feature(:) = string(feat_set{n});
            feats_all_epochs_tb = [feats_all_epochs_tb; feats_tbl];
        end


        %---------------------------------------------------------------------
        % CONNECTIVITY FEATURES
        % (use over all channels but also divide into epochs)
        %---------------------------------------------------------------------
    elseif(strfind(feat_set{n},'connectivity'))

        x_epochs=[]; 
        for c=1:N_channels
            if(c == N_channels)
                [x_epochs(c,:,:), epoch_start_times] = ...
                    overlap_epochs(eeg_data(c,:)',Fs,EPOCH_LENGTH,EPOCH_OVERLAP);
            else
                x_epochs(c,:,:) = overlap_epochs(eeg_data(c,:)',Fs,EPOCH_LENGTH,EPOCH_OVERLAP);
            end
        end
        N_epochs=size(x_epochs,2); 
        
        % B) iterate over epochs:
        feats_epochs=NaN(N_epochs,L_feature);
        x_ep=[];
        
        for e=1:N_epochs
            x_ep=reshape(x_epochs(:,e,:),size(x_epochs,1),size(x_epochs,3));
            
            L_nans=length(find(isnan(x_ep(:))));
            if(100*(L_nans/length(x_ep(:))) < EPOCH_IGNORE_PRC_NANS)
                
                feats_epochs(e,:)=connectivity_features(x_ep,Fs,feat_set{n},[], ...
                                                        ch_labels);
            end
            
        end
        % median over all epochs
        feat_st.(char(feat_set{n}))=nanmedian(feats_epochs, 1);
        
        % if want to return feature estimated over all epochs:
        if(return_feat_epoch)
            % create table with features and start time of epoch:
            fb_names = arrayfun(@(x) ['FB' num2str(x)], 1:size(feats_epochs, 2), 'un', false);
            tb = array2table([epoch_start_times' feats_epochs], ...
                             'VariableNames', ['start_time_sec', fb_names]);
            % add channel:
            tb.channel(:) = NaN;
            
            % convert from wide to long format for frequency bands:
            tb = stack(tb, fb_names, 'newDataVariableName', {'feature_value'}, ...
                       'IndexVariableName', {'freq_band'});
            
            % add feature name and combine:
            tb.feature(:) = string(feat_set{n});
            feats_all_epochs_tb = [feats_all_epochs_tb; tb];
        end
        
        

        %---------------------------------------------------------------------
        % inter-burst interval features
        % (use entire recording but channel-by-channel)
        %---------------------------------------------------------------------
    elseif(strfind(feat_set{n},'IBI_'))
        
        % B) iterate over channels
        feats_channel=NaN(N_channels,L_feature); 
        for c=1:N_channels
            feats_channel(c,:)=IBI_features(eeg_data(c,:)',Fs,feat_set{n});
        end
        % and median over all channels:
        feat_st.(char(feat_set{n}))=nanmedian(feats_channel,1);


    end
end




function [x_epochs, start_times] = overlap_epochs(x, Fs, L_window, overlap, window_type)
%---------------------------------------------------------------------
% overlapping epochs in one matrix
%---------------------------------------------------------------------
if(nargin<4 || isempty(overlap)), overlap=50; end
if(nargin<5 || isempty(window_type)), window_type='rect'; end

x = x(:).';

[L_hop, L_epoch, win_epoch] = gen_epoch_window(overlap, L_window, window_type, Fs);

N = length(x);
N_epochs = ceil( (N - (L_epoch - L_hop)) / L_hop );
if(N_epochs < 1) 
    N_epochs = 1; 
    fprintf('| WARNING: signal length is less than segment length (L_epoch - L_hop).\n');
    fprintf('| Adjust ''EPOCH_LENGTH'' or ''EPOCH_OVERLAP'' in ');
    fprintf('''neural_parameters.m'' file.\n');
end
nw = 0:(L_epoch - 1);
ix = 0:(N - 1);


x_epochs = NaN(N_epochs, L_epoch);
start_times = NaN(1, N_epochs);
for k = 1:N_epochs
    nf = nw + (k - 1) * L_hop;
    % zero-pad if outside x:
    nf = nf(ismember(nf, ix)) + 1;
    i_nf = 1:length(nf);
    start_times(k) = (min(nf) - 1) / Fs;

    x_epochs(k, i_nf) = x(nf) .* win_epoch(i_nf).';
end
